ssrando/ssrando
# Custom model and texture recolour support

Aims:

- Add more structured and granular custom model support
  - Create structure for custom model packs
  - Add ability for users to pick model packs for player model and loftwing model individually
  - Add preview to default player and loftwing models
  - Add ability for model pack creators to add custom previews
  - Add ability for model pack creators to define masks for texture recolouring
- Add masked texture recolour support
  - Add ability to select colours for colour groups, textures with corresponding masks will then be recoloured in the patching process
  - Add masks and colour groups for default player and loftwing models
  - Update previews with texture recolours in realtime
- Link tunic swap option to custom models
  - Add ability for custom model pack creators to disable tunic swap option

Details:

- Adds a new "Cosmetics" tab that contains all the UI for the model pack and colour selection
- Moved tunic swap option to "Cosmetics" tab
- Makes use of my own BRRES parsing library (found in the `brresTools` folder), this library is a WIP and currently contains what is needed to make texture recolouring work- it will be updated in the future
- Currently supports player and loftwing model packs, I plan to support items separately in the future
- Player model packs can optionally contain an `AdditionalArcs` folder which essentially acts as an `arc-replacements` folder for the specific model pack. Primarily intended to support custom UI and items but can support any arcs
- Model pack arcs (including `AdditionalArc` arcs) will take priority over `arc-replacements` arcs, as a result, it is no longer possible to use `Alink.arc` or `Bird_Link.arc` arcs in `arc-replacements` (this is intentional as this feature is meant to be a step towards replacing `arc-replacements` anyway)

Tests needed:

- Verify default player and loftwing models with no texture recolours works as usual
- Verify recoloured player and loftwing models appear as expected in game
- Verify selected custom player and loftwing model packs appear in game (will provide sample model packs soon)
- Verify tunic swap option still works
- Verify randomise, reset, randomise all, and reset all buttons work as expected
- Verify that model pack creation (as detailed in the README in the models folder) works
